pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -373,3 +429,65 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
     return nd_shape;
 
 }
+
+bool    is_same_shape(t_ndarray dest, t_ndarray src)
+{
+    int i = 0;
+
+    if (dest.nd != src.nd)
+        return false;
+    while (i < dest.nd)
+    {
+        if (dest.shape[i] != src.shape[i])
+            return false;
+        ++i;
+    }
+    return true;
+}
+
+int get_shape_product(int64_t *shape, int nd, int max_nd)
+{
+    int product = 1;
+
+    for (int i = nd; i < max_nd; ++i)
+        product *= shape[i];
+    return (product);
+}
+
+int index_position(t_ndarray arr, int index, int nd)
+{
+    if (arr.order == order_c)
+        return index;
+    if (nd == 0)
+        return (0);
+    if (nd == arr.nd)
+        return (index % arr.shape[nd - 1]) * arr.strides[nd - 1] + index_position(arr, index, nd - 1);
+    int true_index = (index / (get_shape_product(arr.shape, nd, arr.nd)));
+    if (true_index >= arr.shape[nd - 1])
+        true_index = true_index % arr.shape[nd - 1];
+    return (true_index * arr.strides[nd - 1] + index_position(arr, index, nd - 1));
+}
+
+void array_copy_data(t_ndarray dest, t_ndarray src, int64_t offset)
+{
+    unsigned char *d = (unsigned char*)dest.raw_data;
+    unsigned char *s = (unsigned char*)src.raw_data;
+
+    if (dest.order == src.order)
+    {
+        if ((dest.order == order_c) || (dest.order == order_f && is_same_shape(dest, src)))
+        {
+            memcpy(d + offset * dest.type_size, s, src.buffer_size);
+            return ;
+        }
+    }
+
+    int64_t index = 0;
+
+    while (index < src.length)
+    {
+        memcpy(d + ((index_position(dest, index, dest.nd) + offset) * dest.type_size),
+               s + (index_position(src, index, src.nd) * src.type_size), src.type_size); // copying element by element

```

Comment:
Do you know what the speed difference is between doing:
```
memcpy(d + ((index_position(dest, index, dest.nd) + offset) * dest.type_size),
               s + (index_position(src, index, src.nd) * src.type_size), src.type_size);
```
vs
```
dest.raw_data[index_position(dest, index, dest.nd) + offset] =
              src.raw_data[index_position(src, index, src.nd) * src.type_size];
```
?

Reply:
No, I need to see if this workes, I assumed that the `[]` operator would reference one byte if it is used with `arr.raw_data`, when we sometimes need to copy 4 bytes (one integer element is 4 bytes)

Reply:
The `void` is not assignable, but I can cast it to `char` or `unsigned char`, which would only reference one byte when used, which would require a `for` loop when trying to copy one element with 8/4 bytes, `memcpy` is faster in this case?

Reply:
Ah you are right. I was hoping we could improve the readability but there's no hope if we don't know the type

Reply:
We do know the type, we can use a switch statemement, there would be a case for every type:
```c
switch (dest.type)
        {
            case nd_int8:
                dest.nd_int8[i_dest + pad] = src.nd_int8[i_src];
                break;
            case nd_int16:
                dest.nd_int16[i_dest + pad] = src.nd_int16[i_src];
                break;
            case nd_int32:
                dest.nd_int32[i_dest + pad] = src.nd_int32[i_src];
                break;
            case nd_int64:
                dest.nd_int64[i_dest + pad] = src.nd_int64[i_src];
                break;
            case nd_float:
                dest.nd_float[i_dest + pad] = src.nd_float[i_src];
                break;
            case nd_double:
                dest.nd_double[i_dest + pad] = src.nd_double[i_src];
                break;
            case nd_bool:
                dest.nd_bool[i_dest + pad] = src.nd_bool[i_src];
                break;
            case nd_cfloat:
                dest.nd_cfloat[i_dest + pad] = src.nd_cfloat[i_src];
                break;
            case nd_cdouble:
                dest.nd_int64[i_dest + pad] = src.nd_cdouble[i_src];
                break;
        }
```
It was originally like this, the assignment operator should be faster than `memcpy`, although this is more readable, I do think the smaller version is better since it is not unreadable
What do you think?

Reply:
Hmmm, readable is good, so is fast, but there's a lot to be said for not repeating things. I don't have a preference without seeing the benchmarks for how big the speed difference is